### DISCUSSION TITLE
stm32_hal -- undefined reference to "SystemCoreClock"

### DISCUSSION THREAD

**Discussion by isdv** [zephyrproject-rtos/zephyr#51677] [Q&A] [answered]
**Describe the bug**
Error compiling "Hello world" example with nucleo_f429zi board and stm32_hal.

I copied the hello_world  to a new directory and installed it in it stm32_hal.

Edited by CMAKELIst.txt:

    cmake_minimum_required(VERSION 3.20.0)

    find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
    project(hello_world)

    target_sources(app PRIVATE src/main.c)
    add_subdirectory(hal_stm32)
    zephyr_library_include_directories(.)

**To Reproduce**
1. mkdir hw; cp -r ~/Devel/SDK/nRFC/zephyr/samples/hello_world/* ./
2. cd hw
4. git clone https://github.com/zephyrproject-rtos/hal_stm32.git
5. west build -b nucleo_f429zi
6. See error
\'../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.bfd: zephyr/libzephyr.a(soc.c.obj): in function 'st_stm32f4_init':
/home/vz/Devel/SDK/nRFC/zephyr/soc/arm/st_stm32/stm32f4/soc.c:51: undefined reference to 'SystemCoreClock'
/home/vz/Devel/tool/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.bfd: zephyr/drivers/clock_control/libdrivers__clock_control.a(clock_stm32_ll_common.c.obj): in function 'get_pll_div_frequency':
/home/vz/Devel/SDK/nRFC/zephyr/drivers/clock_control/clock_stm32_ll_common.c:75: undefined reference to 'SystemCoreClock'
/home/vz/Devel/tool/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.bfd: zephyr/drivers/clock_control/libdrivers__clock_control.a(clock_stm32_ll_common.c.obj): in function 'stm32_clock_control_init':
/home/vz/Devel/SDK/nRFC/zephyr/drivers/clock_control/clock_stm32_ll_common.c:629: undefined reference to 'HAL_RCC_GetSysClockFreq'
/home/vz/Devel/tool/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.bfd: /home/vz/Devel/SDK/nRFC/zephyr/drivers/clock_control/clock_stm32_ll_common.c:637: undefined reference to 'LL_SetFlashLatency'
/home/vz/Devel/tool/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.bfd: /home/vz/Devel/SDK/nRFC/zephyr/drivers/clock_control/clock_stm32_ll_common.c:683: undefined reference to 'LL_SetFlashLatency'
/home/vz/Devel/tool/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.bfd: /home/vz/Devel/SDK/nRFC/zephyr/drivers/clock_control/clock_stm32_ll_common.c:708: undefined reference to 'AHBPrescTable'
/home/vz/Devel/tool/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.bfd: /home/vz/Devel/SDK/nRFC/zephyr/drivers/clock_control/clock_stm32_ll_common.c:708: undefined reference to 'SystemCoreClock'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /tmp/hw/build
\`

**Expected behavior**
Ssuccessful compilation of the project.

**Environment (please complete the following information):**
 - OS: Linux (Fedora 36)
 - Toolchain: NRF Connect SDK v2.1.1/gcc-arm-none-eabi-10-2020-q4-major

**Comment by stephanosio** [comment]
> I copied the hello_world to a new directory and installed it in it stm32_hal.

This is not a valid way to do this.

**Comment by isdv** [comment]
Please suggest the correct way to connect to the project stm32_hal.

**Reply by stephanosio** [reply]
You should not need to do anything special for that.

Simply running `west build -p always -b nucleo_f429zi samples/hello_world` should be enough.

**Comment by isdv** [comment]
west build -b nucleo_f429zi ~/Devel/SDK/nRFC/zephyr/samples/hello_world -p always

`
FAILED: zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj 
ccache /home/vz/Devel/tool/gcc-arm/bin/arm-none-eabi-gcc -DKERNEL -DUSE_PARTITION_MANAGER=0 -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR__=1 -I/home/vz/Devel/SDK/nRFC/zephyr/kernel/include -I/home/vz/Devel/SDK/nRFC/zephyr/arch/arm/include -I/home/vz/Devel/SDK/nRFC/zephyr/include/zephyr -I/home/vz/Devel/SDK/nRFC/zephyr/include -I/tmp/hw2/build/zephyr/include/generated -I/home/vz/Devel/SDK/nRFC/zephyr/soc/arm/st_stm32/stm32f4 -I/home/vz/Devel/SDK/nRFC/zephyr/drivers -I/home/vz/Devel/SDK/nRFC/zephyr/soc/arm/st_stm32/common -I/home/vz/Devel/SDK/nRFC/nrf/include -I/home/vz/Devel/SDK/nRFC/modules/hal/cmsis/CMSIS/Core/Include -isystem /home/vz/Devel/SDK/nRFC/zephyr/lib/libc/minimal/include -isystem /home/vz/Devel/tool/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/include -isystem /home/vz/Devel/tool/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/include-fixed -Os -imacros /tmp/hw2/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee --sysroot=/home/vz/Devel/tool/gcc-arm/arm-none-eabi -imacros /home/vz/Devel/SDK/nRFC/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/vz/Devel/SDK/nRFC/zephyr/samples/hello_world=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/vz/Devel/SDK/nRFC/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/vz/Devel/SDK/nRFC=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -MF zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj.d -o zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -c /home/vz/Devel/SDK/nRFC/zephyr/arch/arm/core/offsets/offsets.c
In file included from /home/vz/Devel/SDK/nRFC/zephyr/include/zephyr/arch/arm/aarch32/cortex_m/cmsis.h:17,
                 from /home/vz/Devel/SDK/nRFC/zephyr/include/zephyr/arch/arm/aarch32/mpu/arm_mpu_v7m.h:11,
                 from /home/vz/Devel/SDK/nRFC/zephyr/include/zephyr/arch/arm/aarch32/mpu/arm_mpu.h:14,
                 from /home/vz/Devel/SDK/nRFC/zephyr/include/zephyr/arch/arm/aarch32/arch.h:210,
                 from /home/vz/Devel/SDK/nRFC/zephyr/include/zephyr/arch/cpu.h:19,
                 from /home/vz/Devel/SDK/nRFC/zephyr/include/zephyr/kernel_includes.h:33,
                 from /home/vz/Devel/SDK/nRFC/zephyr/include/zephyr/kernel.h:17,
                 from /home/vz/Devel/SDK/nRFC/zephyr/arch/arm/core/offsets/offsets_aarch32.c:28,
                 from /home/vz/Devel/SDK/nRFC/zephyr/arch/arm/core/offsets/offsets.c:9:
/home/vz/Devel/SDK/nRFC/zephyr/soc/arm/st_stm32/stm32f4/soc.h:23:10: fatal error: stm32f4xx.h: No such file or directory
   23 | #include <stm32f4xx.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /tmp/hw2/build
`

**Reply by stephanosio** [reply]
It looks like `hal_stm32` module is not pulled in your local environment (maybe because you are using NCS?).

Please try with the upstream Zephyr. This is supposed to work without any modifications.

**Comment by isdv** [comment]
I also copied the file "stm32f429zitx-pinctrl.dtsi " from stm32_hal to the directory zephyr/dts/arm/st/f4   that was required when compiling.

**Reply by stephanosio** [reply]
Please don't do that.

**Comment by isdv** [comment, marked as answer]
Then I get an error:

`zephyr/boards/arm/nucleo_f429zi/nucleo_f429zi.dts:9:10: fatal error: st/f4/stm32f429zitx-pinctrl.dtsi: No such file or directory
    9 | #include <st/f4/stm32f429zitx-pinctrl.dtsi>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
CMake Error at /home/vz/Devel/SDK/nRFC/zephyr/cmake/modules/dts.cmake:189 (message):
  command failed with return code: 1
Call Stack (most recent call first):
  /home/vz/Devel/SDK/nRFC/zephyr/cmake/modules/zephyr_default.cmake:121 (include)
  /home/vz/Devel/SDK/nRFC/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /home/vz/Devel/SDK/nRFC/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  CMakeLists.txt:5 (find_package)
`

**Reply by isdv** [reply]
I installed zephyr instead of nrf connect sdk and everything worked.
